internal/daemon/api_logs.go

```diff
@@ -0,0 +1,318 @@
+		err := encoder.Encode(log)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Follow iterators and output logs as they're written.
+func followLogs(w io.Writer, itsByName map[string]servicelog.Iterator, done <-chan struct{}) {
+	// Start one goroutine per service to listen for and output new logs.
+	writeMutex := &sync.Mutex{}
+	var wg sync.WaitGroup
+	for service, it := range itsByName {
+		wg.Add(1)
+		go func(service string, it servicelog.Iterator) {
+			defer wg.Done()
+			reader := bufio.NewReaderSize(it, bufioReaderSize)
+			fw := &followWriter{w, writeMutex}
+			encoder := json.NewEncoder(fw)
+			encoder.SetEscapeHTML(false)
+			err := iterateLogs(it, reader, done, time.Time{}, func(log logEntry) error {
+				return encoder.Encode(log)
+			})
+			if err != nil {
+				fmt.Fprintf(w, "\nerror writing logs: %v", err)
+				return
+			}
+		}(service, it)
+	}
+
+	// Don't return till client connection is closed and all iterateLogs have
+	// finished. We can't just wait for the done channel here, because the
+	// client connection will close and iterateLogs may still be calling
+	// it.Next() when the caller calls it.Close(), causing a data race.
+	wg.Wait()
+}
+
+// Each log is written as a JSON object followed by a newline (JSON Lines):
+//
+// {"time":"2021-04-23T01:28:52.660Z","service":"redis","message":"redis started up"}
+// {"time":"2021-04-23T01:28:52.798Z","service":"thing","message":"did something"}
+type logEntry struct {
+	Time    time.Time `json:"time"`
+	Service string    `json:"service"`
+	Message string    `json:"message"`
+}
+
+type writeLogFunc func(log logEntry) error
+
+// Iterate through all logs in the given iterator, calling writeLog for each
+// until the iterator is exhausted.
+func iterateLogs(it servicelog.Iterator, br *bufio.Reader, cancel <-chan struct{}, now time.Time, writeLog writeLogFunc) error {
+	var prevTime time.Time
+	var prevService string
+	for it.Next(cancel) {
+		br.Reset(it)
+		for {
+			line, err := br.ReadSlice('\n')
+			if err != nil && !errors.Is(err, io.EOF) {
+				return err
+			}
+			if len(line) == 0 {
+				// No bytes read, we're done with this iteration
+				break
+			}
+
+			log, ok := parseLog(line)
+			if ok {
+				// Output normal log
+				err := writeLog(log)
+				if err != nil {
+					return err
+				}
+				if !now.IsZero() && log.Time.After(now) {
+					// Stop if we're past current time
+					return nil
+				}
+				prevTime = log.Time
+				prevService = log.Service
+			} else if !prevTime.IsZero() {
+				// Output partial log with previous timestamp and service name
+				err := writeLog(logEntry{prevTime, prevService, string(line)})
+				if err != nil {
+					return err
+				}
+			}
+
+			if errors.Is(err, io.EOF) {
+				// EOF reached, break after processing line
+				break
+			}
+		}
+	}
+	return nil
+}
+
+// Parse a log entry of the form "2021-05-20T15:39:12.345Z [service] log message"
+func parseLog(line []byte) (logEntry, bool) {
```

Comment:
Would be great to have this moved to the services log package where the logs / format is defined, and also move the timing format to a common variable there, and then have a test that straps producing and consuming together to make sure we don't get very ugly bugs where we don't respect our own internal inter-dependencies.

Reply:
Good call. I've moved `parseLog` to `servicelog.Parse` and added an iterating `servicelog.Parser`. I've added tests for those, as well as a round-trip test like you're describing that tests the log formatter and log parser together.